{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feb0b531",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "a = ['a','b','c','d']\n",
    "y=[['asd', 2, 3, 4], [5, 6, 'vdf', 8]]\n",
    "\n",
    "z=[[1,2],[3,4],[5,6],[7,8]]\n",
    "\n",
    "x=[[0,1],[2,3]]\n",
    "#x = numpy.array(x)[:,1:]\n",
    "y = [remove[1] for remove in x]\n",
    "print(y)\n",
    "\n",
    "#z = numpy.delete(z, x, axis=0)\n",
    "#print(z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be77eca3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "c = pd.read_csv('test.csv')\n",
    "c = numpy.array(c)[:,2:]\n",
    "print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0e62c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "folder = 'results/fermentation/data/'\n",
    "\n",
    "from joblib import load\n",
    "from pandas import read_csv\n",
    "\n",
    "a = load(folder+'metadata.joblib')\n",
    "print(a)\n",
    "\n",
    "madin = read_csv(folder+'madin.csv')\n",
    "\n",
    "for idx_madin, row_madin in madin.iterrows():\n",
    "\n",
    "    folder = '../data/genomes/'+str(row_madin.taxid)+'/'\n",
    "    if os.path.isdir( folder ) == True:\n",
    "        print(folder)\n",
    "\n",
    "    for file in glob(folder+\"*.fna\"):\n",
    "        \n",
    "        print(file)\n",
    "\n",
    "        if os.path.isfile( file + '.eggnog' ) == False:\n",
    "\n",
    "            if os.path.isfile( file + '.emapper.annotations' ) == True:\n",
    "                #os.system('mv ' + file + '.emapper.annotations ' + file + '.eggnog')\n",
    "                print(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c5bf379",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = '../data/genomes/542/'\n",
    "file = 'GCA_004168325.2_ASM416832v2_genomic.fna.emapper.annotations'\n",
    "if os.path.isfile( folder + file ) == False:\n",
    "    print('a')\n",
    "else:\n",
    "    print('b')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c72fb0a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = 'teste123'\n",
    "a=a.replace('bbb','aaaaaa')\n",
    "print(a)\n",
    "a=a.replace('1','eeeee').replace('2','ffffffffff')\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dac10d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict1 = {'b': 2, 'c': 3, 'a': 1}\n",
    "dict2 = {'a': 1, 'b': 2, 'c': 3}\n",
    "dict3 = {'a': 1, 'b': 2, 'c': 4}\n",
    "\n",
    "# Compare dict1 and dict2 (should return True)\n",
    "print(dict1 == dict2)  # Output: True\n",
    "\n",
    "# Compare dict1 and dict3 (should return False)\n",
    "print(dict1 == dict3)  # Output: False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b383fb3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import numpy\n",
    "\n",
    "folder = 'results/optimum_ph/data/'\n",
    "#folder = 'results/range_salinity/data/'\n",
    "\n",
    "#y = pandas.read_csv(folder + 'metadata.csv')\n",
    "#y = numpy.array(y)[:,6:] # slicing to exclude first 6 columns (files name and taxonomy) and leave only phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "058df69b",
   "metadata": {},
   "outputs": [],
   "source": [
    "type(y[0][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02725ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "y = numpy.hsplit(y, len(y[0]))[0]\n",
    "y = numpy.ravel(y)\n",
    "y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54a8b370",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4588"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy\n",
    "from joblib import load\n",
    "\n",
    "phen = 'optimum_ph'\n",
    "folder = 'results/'+phen+'/data/'\n",
    "\n",
    "\n",
    "OG = load( folder + 'OGColumns.joblib', mmap_mode='r+' )\n",
    "len(OG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "08181dd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = load( folder + 'data.joblib', mmap_mode='r+' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8de3b5d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5868\n",
      "6290\n",
      "6572\n",
      "7040\n",
      "8238\n",
      "8715\n",
      "9019\n",
      "9185\n",
      "14893\n",
      "14949\n",
      "15624\n",
      "15993\n",
      "16761\n",
      "16383\n",
      "16662\n",
      "17720\n",
      "17761\n",
      "19307\n",
      "19602\n",
      "19151\n",
      "20202\n",
      "20391\n",
      "20941\n",
      "21100\n",
      "21521\n",
      "11248\n",
      "21763\n",
      "22807\n",
      "23520\n",
      "23604\n",
      "24237\n",
      "24812\n",
      "24916\n",
      "13670\n",
      "26408\n",
      "26444\n",
      "13971\n",
      "26039\n",
      "26358\n",
      "27570\n",
      "27441\n",
      "28403\n",
      "28168\n",
      "28956\n",
      "29962\n",
      "29982\n",
      "31553\n",
      "32116\n",
      "32801\n",
      "33198\n",
      "34148\n",
      "34277\n",
      "15935\n",
      "34817\n",
      "34979\n",
      "34993\n",
      "35457\n",
      "36205\n",
      "35700\n",
      "35858\n",
      "36270\n",
      "37094\n",
      "36317\n",
      "37916\n",
      "37973\n",
      "38064\n",
      "37439\n",
      "38429\n",
      "38180\n",
      "38251\n",
      "39584\n",
      "39283\n",
      "19265\n",
      "40640\n",
      "41387\n",
      "40459\n",
      "41756\n",
      "40967\n",
      "41063\n",
      "42599\n",
      "41789\n",
      "42959\n",
      "43024\n",
      "43440\n",
      "44657\n",
      "44529\n",
      "45907\n",
      "46366\n",
      "46481\n",
      "46805\n",
      "48243\n",
      "48458\n",
      "49182\n",
      "49691\n",
      "50364\n",
      "50448\n",
      "50569\n",
      "50657\n",
      "50220\n",
      "51102\n",
      "51197\n",
      "50670\n",
      "51582\n",
      "51722\n",
      "51792\n",
      "52068\n",
      "51567\n",
      "52151\n",
      "52995\n",
      "53084\n",
      "53243\n",
      "53733\n",
      "53764\n",
      "53844\n",
      "54058\n",
      "54171\n",
      "54225\n",
      "54279\n",
      "54354\n",
      "54562\n",
      "55363\n",
      "55446\n",
      "55505\n",
      "55677\n",
      "23195\n",
      "56003\n",
      "55430\n",
      "56641\n",
      "56678\n",
      "57028\n",
      "57337\n",
      "57692\n",
      "57739\n",
      "57910\n",
      "57953\n",
      "58002\n",
      "58499\n",
      "58641\n",
      "59091\n",
      "59108\n",
      "59267\n",
      "59533\n",
      "59720\n",
      "59759\n",
      "59884\n",
      "60034\n",
      "60717\n",
      "61082\n",
      "61239\n",
      "61393\n",
      "61570\n",
      "61606\n",
      "61748\n",
      "61118\n",
      "62067\n",
      "62201\n",
      "62243\n",
      "62321\n",
      "62455\n",
      "62708\n",
      "62922\n",
      "63192\n",
      "62578\n",
      "63414\n",
      "63676\n",
      "63717\n",
      "63788\n",
      "63075\n",
      "64089\n",
      "64216\n",
      "64506\n",
      "64549\n",
      "64895\n",
      "64942\n",
      "65120\n",
      "65480\n",
      "65568\n",
      "65708\n",
      "65892\n",
      "65920\n",
      "65994\n",
      "65385\n",
      "66176\n",
      "66211\n",
      "66504\n",
      "66608\n",
      "66733\n",
      "66854\n",
      "66946\n",
      "66995\n",
      "67257\n",
      "67388\n",
      "67658\n",
      "67681\n",
      "67784\n",
      "67925\n",
      "68026\n",
      "68293\n",
      "68314\n",
      "67761\n",
      "68556\n",
      "68589\n",
      "69109\n",
      "69160\n",
      "69175\n",
      "69190\n",
      "69349\n",
      "69425\n",
      "69663\n",
      "69397\n",
      "28114\n",
      "70336\n",
      "70370\n",
      "70579\n",
      "70690\n",
      "70885\n",
      "70909\n",
      "70940\n",
      "71175\n",
      "71282\n",
      "71428\n",
      "71712\n",
      "71879\n",
      "72009\n",
      "72034\n",
      "72203\n",
      "72295\n",
      "72439\n",
      "72615\n",
      "72655\n",
      "72700\n",
      "73009\n",
      "73108\n",
      "72853\n",
      "73560\n",
      "73077\n",
      "73718\n",
      "74179\n",
      "74219\n",
      "73776\n",
      "74498\n",
      "74516\n",
      "74579\n",
      "74651\n",
      "74667\n",
      "74685\n",
      "75090\n",
      "75213\n",
      "75314\n",
      "75551\n",
      "75587\n",
      "75634\n"
     ]
    }
   ],
   "source": [
    "for y in x:\n",
    "    if len(y) != 4588:\n",
    "        print(len(y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e883ebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas import DataFrame\n",
    "\n",
    "OGs = ['asf','fasf']\n",
    "\n",
    "a = DataFrame(columns=['Ortholog Group','Gain']).set_index('Ortholog Group')\n",
    "temp = DataFrame(index='Ortholog Group',data={'Ortholog Group': OGs, 'Gain': estimator.feature_importances_})\n",
    "temp = temp[temp.Gain != 0]\n",
    "gain_importance = gain_importance.add(temp.set_index('Ortholog Group'), fill_value=0).reset_index()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
